File: golang/327.区间和的个数.go
// 327.区间和的个数.go
// https://leetcode.cn/problems/count-of-range-sum/description/
// lang=golang

// 解法: 力扣官方题解

package main

import "fmt"

func countRangeSum(nums []int, lower, upper int) int {
    var mergeCount func([]int) int
    mergeCount = func(arr []int) int {
        n := len(arr)
        if n <= 1 {
            return 0
        }

        n1 := append([]int(nil), arr[:n/2]...)
        n2 := append([]int(nil), arr[n/2:]...)
        cnt := mergeCount(n1) + mergeCount(n2) // 递归完毕后，n1 和 n2 均为有序

        // 统计下标对的数量
        l, r := 0, 0
        for _, v := range n1 {
            for l < len(n2) && n2[l]-v < lower {
                l++
            }
            for r < len(n2) && n2[r]-v <= upper {
                r++
            }
            cnt += r - l
        }

        // n1 和 n2 归并填入 arr
        p1, p2 := 0, 0
        for i := range arr {
            if p1 < len(n1) && (p2 == len(n2) || n1[p1] <= n2[p2]) {
                arr[i] = n1[p1]
                p1++
            } else {
                arr[i] = n2[p2]
                p2++
            }
        }
        return cnt
    }

    prefixSum := make([]int, len(nums)+1)
    for i, v := range nums {
        prefixSum[i+1] = prefixSum[i] + v
    }
    return mergeCount(prefixSum)
}

// 作者：力扣官方题解
// 链接：https://leetcode.cn/problems/count-of-range-sum/solutions/476038/qu-jian-he-de-ge-shu-by-leetcode-solution/
// 来源：力扣（LeetCode）
// 著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。


func main() {
	f := countRangeSum
	fmt.Println(f([]int{-2, 5, -1}, -2, 2))
	fmt.Println(f([]int{0}, 0, 0))
	// 输出: 3
	// 输出: 1
}

File: python/327.区间和的个数.py
# 327.区间和的个数.py
# https://leetcode.cn/problems/count-of-range-sum/description/
# lang=python
from itertools import accumulate
from typing import List


# 解法1: mergesort_模板

class Solution:
    def countRangeSum(self, nums: List[int], lower: int, upper: int) -> int:
        self.lower = lower
        self.upper = upper
        self.res = 0
        n = len(nums)
        prefixsum = [0 for _ in range(n + 1)]
        for i in range(n):
            prefixsum[i + 1] = prefixsum[i] + nums[i]
        
        self.mergesort(prefixsum, 0, len(prefixsum) - 1)
        return self.res
    
    def mergesort(self, nums: List[int], L: int, R: int) -> None:
        if L < R:
            mid = L + (R - L) // 2
            self.mergesort(nums, L, mid)
            self.mergesort(nums, mid + 1, R)
            self.merge(nums, L, mid, R)
    
    def merge(self, nums: List[int], L: int, mid: int, R: int) -> None:
        #########################################
        ## 套用标准的归并排序   本题需要单独计算的地方
        ii, jj, kk = L, mid + 1, mid + 1
        while ii <= mid:
            while jj <= R and nums[jj] - nums[ii] < self.lower:
                jj += 1
            while kk <= R and nums[kk] - nums[ii] <= self.upper:
                kk += 1
            self.res += (kk - jj)
            ii += 1
        #########################################
        i, j = L, mid + 1
        tmp = []
        while i <= mid and j <= R:
            if nums[i] <= nums[j]:
                tmp.append(nums[i])
                i += 1
            else:
                tmp.append(nums[j])
                j += 1
        if i <= mid:
            tmp += nums[i: mid + 1]
        if j <= R:
            tmp += nums[j: R + 1]
        # for i in range(len(tmp)):
        #     nums[L + i] = tmp[i]
        nums[L:R + 1] = tmp


# 作者：XingHe
# 链接：https://leetcode.cn/problems/count-of-range-sum/solutions/1256482/cpython3-0er-fen-1gui-bing-pai-xu-2shu-z-9dc4/
# 来源：力扣（LeetCode）
# 著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。

####################################################################################

# 解法2: 官解写法

class Solution1:
    def countRangeSum(self, nums: List[int], lower: int, upper: int) -> int:
        prefixSum = list(accumulate(nums, initial=0))
        
        def mergeCount(arr: List[int]) -> int:
            n = len(arr)
            if n <= 1:
                return 0
            
            n1, n2 = arr[:n // 2], arr[n // 2:]
            cnt = mergeCount(n1) + mergeCount(n2)
            l, r = 0, 0
            for v in n1:
                while l < len(n2) and n2[l] < v + lower:
                    l += 1
                while r < len(n2) and n2[r] < v + upper + 1:
                    r += 1
                cnt += r - l
            
            p1, p2 = 0, 0
            for i in range(len(arr)):
                if p1 < len(n1) and (p2 == len(n2) or n1[p1] <= n2[p2]):
                    arr[i] = n1[p1]
                    p1 += 1
                else:
                    arr[i] = n2[p2]
                    p2 += 1
            return cnt
        
        return mergeCount(prefixSum)


if __name__ == "__main__":
    import time
    
    t_0 = time.time()
    
    s = Solution()
    # s = Solution1()
    func = getattr(s, "countRangeSum")
    print(func)
    
    print(func(nums=[-2, 5, -1], lower=-2, upper=2))
    # 输出：3
    print(func(nums=[0], lower=0, upper=0))
    # 输出：1
    
    t_1 = time.time()
    print(t_1 - t_0)
